### ISSUE TITLE
Question on maskings

### ISSUE THREAD

**Issue by duyvuleo** [Kyubyong/transformer#16]
Hi @Kyubyong,

Can you help explain a bit on the following masking codes (the Key Masking and Query Masking)? Why we need them? We only need the causality, right?

```
# Key Masking
        key_masks = tf.sign(tf.abs(tf.reduce_sum(keys, axis=-1))) # (N, T_k)
        key_masks = tf.tile(key_masks, [num_heads, 1]) # (h*N, T_k)
        key_masks = tf.tile(tf.expand_dims(key_masks, 1), [1, tf.shape(queries)[1], 1]) # (h*N, T_q, T_k)
        
        paddings = tf.ones_like(outputs)*(-2**32+1)
        outputs = tf.where(tf.equal(key_masks, 0), paddings, outputs) # (h*N, T_q, T_k)
```

```
# Query Masking
        query_masks = tf.sign(tf.abs(tf.reduce_sum(queries, axis=-1))) # (N, T_q)
        query_masks = tf.tile(query_masks, [num_heads, 1]) # (h*N, T_q)
        query_masks = tf.tile(tf.expand_dims(query_masks, -1), [1, 1, tf.shape(keys)[1]]) # (h*N, T_q, T_k)
        outputs *= query_masks # broadcasting. (N, T_q, C)
```

Thanks!

**Comment by shrshore** [comment]
Also wondering what's the purpose of these query masking and key masking. Can anyone help? Thanks very much. @duyvuleo @Kyubyong

**Comment by shrshore** [comment]
@duyvuleo Hi, the query masking and key masking are utilized to filtered out the attention allocated to `padding' in the sequence.

**Comment by duyvuleo** [comment]
@shrshore : Thanks. I now understood it.

**Comment by YFwang1992** [comment]
Hi @duyvuleo @duyvuleo:
    please help to understand maskings . i want to know masking make 0 to (-2**32+1) ,why do this?
0 can't allocated `padding' in the sequence?

**Comment by gitfourteen** [comment]
@haoransh When you pass **inputs** as an argument to function **positional_encoding**, yes, the **inputs** consists of padding info. However, inside of **positional_encoding**, we just extract the shape info of **inputs** , without padding info. That means, the zero embedding vector of lookup table of **positional_encoding** is NOT the same thing as the lookup table of the word embedding.

**_if hp.sinusoid_**, this would result in the position-encoded padding embedding-vector non-zeros and the mask simply does not fulfill its original purpose.

else, if zero_pad = True, the mask works, since positional encoding and word embedding use the same **embedding** function, except _vocab_size_

**Comment by haoransh** [comment]
@gitfourteen @jiangxinyang227  Yes, this issue has been posted here https://github.com/Kyubyong/transformer/issues/33 before.

So this repo can only serve as a toy example, not the same as the original implementation in tensor2tensor. Also if you are interested, you can also refer to another tensorflow implementation [here](https://github.com/asyml/texar/tree/master/examples/transformer), which is the same as the original implementation but much easier to follow than tensor2tensor.

**Comment by HighWay-To-Hell** [comment]
Also, see [#3](https://github.com/Kyubyong/transformer/issues/3)
